File: src/services/rag/rerankService.js
/**
 * Re-ranking Service
 * Uses a cross-encoder model to re-rank retrieved chunks for better relevance.
 * 
 * Cross-encoders score query-document pairs directly, providing more accurate
 * relevance scores than bi-encoder (embedding) similarity.
 */
const { pipeline, env } = require('@xenova/transformers');

// Configure transformers.js to use local cache
env.cacheDir = './.cache/transformers';
env.allowRemoteModels = true;

let rerankPipeline = null;
let isLoading = false;
let loadPromise = null;

// Default model - a lightweight cross-encoder trained on MS MARCO
const DEFAULT_MODEL = 'Xenova/ms-marco-MiniLM-L-6-v2';

/**
 * Initialize the cross-encoder pipeline (lazy loading).
 * @returns {Promise<Function>}
 */
async function getReranker() {
  if (rerankPipeline) return rerankPipeline;
  
  if (isLoading) {
    // Wait for the existing load to complete
    return loadPromise;
  }

  isLoading = true;
  console.log('[Reranker] Loading cross-encoder model (first run may take a minute)...');
  
  loadPromise = pipeline('text-classification', DEFAULT_MODEL, {
    quantized: true // Use quantized model for faster inference
  });

  try {
    rerankPipeline = await loadPromise;
    console.log('[Reranker] Model loaded successfully');
    return rerankPipeline;
  } catch (err) {
    isLoading = false;
    loadPromise = null;
    throw new Error(`Failed to load reranker model: ${err.message}`);
  }
}

/**
 * Re-rank chunks using cross-encoder scores.
 * @param {string} query - The search query
 * @param {Object[]} chunks - Array of { text, metadata, similarityScore, ... }
 * @param {Object} opts - { topN: number of results to return (default: all) }
 * @returns {Promise<Object[]>} Re-ranked chunks with crossEncoderScore added
 */
async function rerankChunks(query, chunks, opts = {}) {
  if (!chunks || chunks.length === 0) {
    return [];
  }

  // If only 1 chunk, no need to re-rank
  if (chunks.length === 1) {
    return chunks.map(c => ({ ...c, crossEncoderScore: 1.0 }));
  }

  const reranker = await getReranker();
  const startTime = Date.now();

  // Build query-document pairs for the cross-encoder
  const pairs = chunks.map(chunk => ({
    text: `${query} [SEP] ${chunk.text.substring(0, 512)}` // Truncate to model max length
  }));

  // Score all pairs
  const scores = [];
  for (let i = 0; i < pairs.length; i++) {
    try {
      const result = await reranker(pairs[i].text);
      // The model outputs a relevance score
      const score = result[0]?.score || 0;
      scores.push({ index: i, score });
    } catch (err) {
      console.warn(`[Reranker] Failed to score chunk ${i}: ${err.message}`);
      scores.push({ index: i, score: 0 });
    }
  }

  // Sort by cross-encoder score (descending)
  scores.sort((a, b) => b.score - a.score);

  // Rebuild chunks array with cross-encoder scores
  const rerankedChunks = scores.map(s => ({
    ...chunks[s.index],
    crossEncoderScore: parseFloat(s.score.toFixed(4)),
    originalRank: s.index + 1
  }));

  // Optionally limit to topN
  const topN = opts.topN || rerankedChunks.length;
  const result = rerankedChunks.slice(0, topN);

  const elapsed = Date.now() - startTime;
  console.log(`[Reranker] Re-ranked ${chunks.length} chunks in ${elapsed}ms`);

  return result;
}

/**
 * Calculate a combined score from embeddings and cross-encoder.
 * @param {number} embeddingScore - Similarity from vector search (0-1)
 * @param {number} crossEncoderScore - Score from cross-encoder (typically 0-1)
 * @param {number} embeddingWeight - Weight for embedding score (default 0.3)
 * @returns {number} Combined score
 */
function combineScores(embeddingScore, crossEncoderScore, embeddingWeight = 0.3) {
  const ceWeight = 1 - embeddingWeight;
  return (embeddingScore * embeddingWeight) + (crossEncoderScore * ceWeight);
}

/**
 * Check if the reranker is ready (model loaded).
 */
function isReady() {
  return rerankPipeline !== null;
}

/**
 * Preload the model (useful at startup).
 */
async function preload() {
  await getReranker();
}

module.exports = {
  rerankChunks,
  combineScores,
  isReady,
  preload
};

File: app/services/rag/rerank_service.py
"""
Re-ranking Service
Keyword-overlap re-ranker that combines vector similarity scores
with BM25-like term-frequency relevance for better retrieval quality.

This lightweight approach avoids heavy ML model dependencies while
still providing meaningful re-ranking of candidate chunks.
"""
import math
import re
from collections import Counter
from typing import Dict, List, Optional


def rerank_chunks(
    query: str,
    chunks: List[Dict],
    top_n: Optional[int] = None,
    vector_weight: float = 0.5,
    keyword_weight: float = 0.5,
) -> List[Dict]:
    """
    Re-rank chunks using a combined score of vector similarity and keyword overlap.

    Args:
        query: The search query.
        chunks: List of dicts with 'text', 'metadata', 'similarity_score'.
        top_n: Return only the top N results (default: all).
        vector_weight: Weight for vector similarity [0-1].
        keyword_weight: Weight for keyword score [0-1].

    Returns:
        Re-ranked list of chunks with 'rerank_score' and 'original_rank' added.
    """
    if not chunks:
        return []
    if len(chunks) == 1:
        chunks[0]["rerank_score"] = chunks[0].get("similarity_score", 1.0)
        chunks[0]["original_rank"] = 1
        return chunks

    query_terms = _tokenize(query)
    if not query_terms:
        return chunks

    # Compute IDF across the candidate set
    doc_freq: Counter = Counter()
    for chunk in chunks:
        tokens = set(_tokenize(chunk.get("text", "")))
        for t in tokens:
            doc_freq[t] += 1
    n_docs = len(chunks)

    scored = []
    for rank, chunk in enumerate(chunks):
        # Vector similarity (already 0-1 from ChromaDB)
        vec_score = chunk.get("similarity_score", 0.0)

        # BM25-like keyword score
        kw_score = _bm25_score(query_terms, chunk.get("text", ""), doc_freq, n_docs)

        combined = (vector_weight * vec_score) + (keyword_weight * kw_score)
        scored.append({
            **chunk,
            "rerank_score": round(combined, 4),
            "keyword_score": round(kw_score, 4),
            "original_rank": rank + 1,
        })

    # Sort by combined score descending
    scored.sort(key=lambda x: x["rerank_score"], reverse=True)

    if top_n and top_n < len(scored):
        scored = scored[:top_n]

    return scored


# ─── Helpers ──────────────────────────────────────────────

def _tokenize(text: str) -> List[str]:
    """Simple whitespace + punctuation tokenizer with lowercasing."""
    return [w.lower() for w in re.findall(r"\b\w+\b", text) if len(w) > 1]


def _bm25_score(
    query_terms: List[str],
    doc_text: str,
    doc_freq: Counter,
    n_docs: int,
    k1: float = 1.5,
    b: float = 0.75,
    avg_dl: float = 200.0,
) -> float:
    """
    Simplified BM25 relevance score.
    """
    doc_tokens = _tokenize(doc_text)
    dl = len(doc_tokens)
    if dl == 0:
        return 0.0

    tf_counter = Counter(doc_tokens)
    score = 0.0

    for term in query_terms:
        tf = tf_counter.get(term, 0)
        df = doc_freq.get(term, 0)
        if tf == 0 or df == 0:
            continue
        idf = math.log((n_docs - df + 0.5) / (df + 0.5) + 1.0)
        tf_norm = (tf * (k1 + 1)) / (tf + k1 * (1 - b + b * (dl / avg_dl)))
        score += idf * tf_norm

    # Normalize to 0-1 range (approximate)
    max_possible = len(query_terms) * 3.0  # rough upper bound
    return min(score / max_possible, 1.0) if max_possible > 0 else 0.0
